### PR TITLE
Offline: Plumb clientSequenceNumber from loader submit APIs to PendingStateManager again

### PR DESCRIPTION
**Pull request by markfields**
## Description

Fixes [AB#8282](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/8282).  It's a partial revert of #13994.

The Container's various "submit" functions all return a client sequence number (CSN) which is assigned sequentially when the loader layer sends the message to the ordering service.  For a batch, the CSN returned is the CSN of the last message in the batch, including chunks sent during chunking (if applicable).

We take this and convert it to the CSN of the start of the back, which will be used to form a new Batch ID for detecting duplicate changes across container forks, e.g. when improperly rehydrating from a serialized container.

### CODE REVIEW

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts [hunk 1]
```diff
@@ -380,10 +387,10 @@ export class Outbox {
 				0x5a6 /* Compression should not have happened if the loader does not support it */,
 			);
 
-			this.params.legacySendBatchFn(batch);
+			clientSequenceNumber = this.params.legacySendBatchFn(batch);
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/opLifecycle/outbox.ts:390]: Do we know for sure this returns clientSequenceNumber?

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts [hunk 2]
```diff
@@ -380,10 +387,10 @@ export class Outbox {
...
 		} else {
 			assert(batch.referenceSequenceNumber !== undefined, 0x58e /* Batch must not be empty */);
-			this.params.submitBatchFn(
+			clientSequenceNumber = this.params.submitBatchFn(
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/opLifecycle/outbox.ts:393]: Same here

File: packages/runtime/container-runtime/src/pendingStateManager.ts [hunk 3]
```diff
@@ -270,6 +275,18 @@ export class PendingStateManager implements IDisposable {
 
 		const messageContent = buildPendingMessageContent(message);
 
+		// Client Sequence Number should match per batch
+		// The server sequences messages in the order received, just like the pending message queue
+		if (this.pendingBatchBeginMessage?.clientSequenceNumber !== pendingMessage.batchStartCsn) {
+			this.logger?.sendErrorEvent({
+				eventName: "PendingClientSequenceNumberMismatch",
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/pendingStateManager.ts:282]: Looks like some test is hitting this error

Comment (markfields) [reply packages/runtime/container-runtime/src/pendingStateManager.ts:284]: Yeah had 2 bugs. One around grouped batching and the other around single-message batches of any kind.

I'm going to have to remove this for now, the fix for the grouped batching one deserves its own PR (and adding this check is not a priority - it was a learning opportunity for me)

Comment (markfields) [reply packages/runtime/container-runtime/src/pendingStateManager.ts:284]: FYI @dannimad

File: packages/runtime/container-runtime/src/opLifecycle/outbox.ts (hunk 1 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/opLifecycle/outbox.ts:390]: Yes - it was like this before #13994 

And for fun I browsed back via git blame... 2019 was the last time I can't immediately prove it returns a CSN.
https://github.com/microsoft/FluidFramework/pull/839
https://github.com/microsoft/FluidFramework/commit/05a5df0106238896ac2913d2bae356ccd3cde4e3

File: packages/runtime/container-runtime/src/pendingStateManager.ts [hunk 4]
```diff
@@ -187,27 +189,30 @@ export class PendingStateManager implements IDisposable {
...
-			content,
-			localOpMetadata,
-			opMetadata,
-		};
-
-		this.pendingMessages.push(pendingMessage);
+	public onFlushBatch(batch: BatchMessage[], clientSequenceNumber: number | undefined) {
+		for (const message of batch) {
+			const {
+				contents: content = "",
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/pendingStateManager.ts:201]: Why do we need this "" here? This seems odd when considering we didn't have it before in the previous `onSubmitMessage`

File: packages/runtime/container-runtime/src/pendingStateManager.ts [hunk 5]
```diff
@@ -187,27 +189,30 @@ export class PendingStateManager implements IDisposable {
...
-	 * @param type - The container message type.
-	 * @param content - The message content.
-	 * @param localOpMetadata - The local metadata associated with the message.
+	 * The given batch has been flushed, and needs to be tracked locally until the corresponding
+	 * acks are processed, to ensure it is successfully sent.
+	 * @param batch - The batch that was flushed
+	 * @param clientSequenceNumber - The CSN of the first message in the batch,
+	 * or undefined if the batch was not yet sent (e.g. by the time we flushed we lost the connection)
 	 */
-	public onSubmitMessage(
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/pendingStateManager.ts:196]: Maybe this is me overthinking things, (non-blocking) why don't we just keep `onSubmitMessage` and call `onSubmitMessage` during `onFlushBatch` or even convert `onSubmitMessage` to `private`?

Comment (markfields) [reply packages/runtime/container-runtime/src/pendingStateManager.ts:196]: Direction is to have PSM deal in while batches not individual messages

File: packages/runtime/container-runtime/src/pendingStateManager.ts (hunk 4 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/pendingStateManager.ts:201]: See old outbox line 404, it was doing ! 

I have a change stashed that fixes the types here to align.

Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/pendingStateManager.ts:201]: Maybe make this an assert, not sure

**Review by tyler-cai-microsoft — APPROVED** [review]
LGTM

**Review by dannimad — APPROVED** [review]
LGTM as well. Just curious about why you moved some batch logic to the pending state manager. It looks to me that batch logic could have been left on outbox and psm keep talking in ops. Is it some kind of intermediate step towards changing the whole pending ops processing?

**Pull request merged**